arwen/cryptoapi/cryptoei.go

```diff
@@ -287,3 +372,607 @@ func v1_4_verifySecp256k1(
+
+	result, xPubKeyGK, yPubKeyGK, err := elliptic.GenerateKey(ec, rand.Reader)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return int32(len(result))
+	}
+
+	err = runtime.MemStore(resultOffset, result)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return int32(len(result))
+	}
+
+	xPubKey.Set(xPubKeyGK)
+	yPubKey.Set(yPubKeyGK)
+	return 0
+}
+
+//export v1_4_createEC
+func v1_4_createEC(context unsafe.Pointer, dataOffset int32, dataLength int32) int32 {
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.EllipticCurveNew
+	metering.UseGas(gasToUse)
+
+	data, err := runtime.MemLoad(dataOffset, dataLength)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return -1
+	}
+	curveChoice := string(data[:])
+	switch curveChoice {
+	case "p224":
+		{
+			curveParams := elliptic.P224().Params()
+			return managedType.PutEllipticCurve(curveParams)
+		}
+	case "p256":
+		{
+			curveParams := elliptic.P256().Params()
+			return managedType.PutEllipticCurve(curveParams)
+		}
+	case "p384":
+		{
+			curveParams := elliptic.P384().Params()
+			return managedType.PutEllipticCurve(curveParams)
+		}
+	case "p521":
+		{
+			curveParams := elliptic.P521().Params()
+			return managedType.PutEllipticCurve(curveParams)
+		}
+	}
+	return -1
+}
+
+//export v1_4_getCurveLengthEC
+func v1_4_getCurveLengthEC(context unsafe.Pointer, ecHandle int32) int32 {
+	managedType := arwen.GetManagedTypesContext(context)
+	metering := arwen.GetMeteringContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+
+	gasToUse := metering.GasSchedule().BigIntAPICost.BigIntGetInt64
+	metering.UseGas(gasToUse)
+
+	ecLength := managedType.GetEllipticCurveSizeOfField(ecHandle)
+	if ecLength == -1 {
+		arwen.WithFault(arwen.ErrNoEllipticCurveUnderThisHandle, context, runtime.BigIntAPIErrorShouldFailExecution())
+	}
+
+	return ecLength
+}
+
+//export v1_4_getPrivKeyByteLengthEC
+func v1_4_getPrivKeyByteLengthEC(context unsafe.Pointer, ecHandle int32) int32 {
+	managedType := arwen.GetManagedTypesContext(context)
+	metering := arwen.GetMeteringContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+
+	gasToUse := metering.GasSchedule().BigIntAPICost.BigIntGetInt64
+	metering.UseGas(gasToUse)
+
+	byteLength := managedType.GetPrivateKeyByteLengthEC(ecHandle)
+	if byteLength == -1 {
+		arwen.WithFault(arwen.ErrNoEllipticCurveUnderThisHandle, context, runtime.BigIntAPIErrorShouldFailExecution())
+	}
+
+	return byteLength
+}
+
+//export v1_4_ellipticCurveGetValues
+func v1_4_ellipticCurveGetValues(context unsafe.Pointer, ecHandle int32, fieldOrderHandle int32, basePointOrderHandle int32, eqConstantHandle int32, xBasePointHandle int32, yBasePointHandle int32) int32 {
```

Comment:
Name of function is somewhat inconsistent with the rest. Maybe `getParametersEC()`? But this is optional.